Added MultiFunctionCorrelationModel and supporting classes.

johnpdpkarp opened the pull request:
Added SPDCorrelationFunction base class and three derived classes :  FourParameterCorrelationFunction, LinearDecayCorrelationFunction, and DampedCosineCorrelationFunction.

Modified FourParameterCorrelationModel and LinearDecayCorrelationModel classes to use new correlation functions for consistency.

Added new MultiFunctionCorrelationModel class which can utilize multiple (different) derived instances of SPDCorrelationFunction .

RubeRad commented:
I thought this already existed (that 4param and lineardecay inherited from a base SPD class/interface), but I guess what you're doing here is making up for the fact that the SPD base didn't exist, it was just 4param/LD out there as two unrelated classes?

Why are there .o files included in the commit?

sminster commented on FourParameterCorrelationModel.cpp:
```diff
@@ -22,20 +22,24 @@
+#include "FourParameterCorrelationFunction.h"
 #include "Error.h"
 
+#ifdef IRIXN32
```
Probably don't need to worry about SGI (IRIX) support anymore.

sminster commented on FourParameterCorrelationModel.cpp:
```diff
@@ -22,20 +22,24 @@
 //     20-Dec-2012   JPK      Fixed bug in constructor.
 //     18-Feb-2013   JPK      Removed getNumCorrelationParameterGroups,
 //                            now provided on base class.
-//     03-Sep-2021   SCM      Removed IRIX support.
```
You seem to have removed this history entry (and likely the associated changes).

sminster commented on LinearDecayCorrelationFunction.cpp:
```diff
@@ -0,0 +1,189 @@
+   for(size_t s = 1; s < size; ++s)
+   {
+      const double corr = params.theInitialCorrsPerSegment[s];
+      const double time = params.theTimesPerSegment[s];
```
There is also a system call named `time()`, so naming this variable `time` could be confusing.

johnpdpkarp replied:
Renamed  "corr" to "currCorr" and "time" to "currTime"

sminster commented on LinearDecayCorrelationFunction.h:
```diff
@@ -0,0 +1,127 @@
+   LinearDecayCorrelationFunction(const std::vector<double>& initialCorrsPerSegment,
+                                  const std::vector<double>& timesPerSegment);
+   //> This constructor instantiates from the provided parameters.
+   //<
```
Inconsistent indentation with other documentation comments in this header.  Ideally, you should indent these to 6 spaces.

sminster commented on LinearDecayCorrelationFunction.h:
```diff
@@ -0,0 +1,127 @@
+   //< 
+protected:
+
+      LinearDecayCorrelationModel::Parameters theCorrParams;
```
Incorrect indentation on this data member.

johnpdpkarp replied:
Fixed

sminster commented on LinearDecayCorrelationModel.cpp:
```diff
@@ -21,18 +21,22 @@
 //     20-Dec-2012   JPK      Fixed bug in constructor.
 //     18-Feb-2013   JPK      Removed getNumCorrelationParameterGroups,
 //                            now provided on base class.
-//     03-Sep-2021   SCM      Removed IRIX support.
```
Undo reversion of this change (like the `#ifdef` below)

johnpdpkarp replied:
Fixed

sminster reviewed (changes requested):
Some minor things to fix or improve.  Also, remove the .o files.

sminster commented on DampedCosineCorrelationFunction.cpp:
```diff
@@ -0,0 +1,122 @@
+DampedCosineCorrelationFunction::getCorrelationCoefficient(double deltaTime) const
+{
+   // compute the value of the correlation coefficient
+   constexpr double pi = 3.14159265358979323846;
```
You may have to add an include for `<cmath>` and add `-D_USE_MATH_DEFINES` in `Makefile.win32`

johnpdpkarp replied:
I'm not convinced this is the right thing to do.  While I realize many compilers do include M_PI in <cmath>, it isn't actually part of the language.  Adding that compile flag in Makefile.win32 may make it easy for someone to build the csm API library standalaone, but software toolkits (such as MSP) build the csm API library themselves in their own way... Making the change in Makefile.win32 won't have any impact on them...       PI is not a defined constant for C++ until C++20 when it will be available as a mathematical constant... and unfortunately years away from C++20.

RubeRad replied:
You could always 
#ifndef M_PI
#define M_PI 3.14...;
#endif

sminster replied:
@johnpdpkarp Modifying `Makefile.win32` would be OK because this constant is only used in the `.cpp` file, not a header file.  Or you could do like @RubeRad  suggests, but I think getting `M_PI` from the compiler is probably better.

roseej replied:
Better to have a csm constant for PI I think.

johnpdpkarp commented on Makefile:
```diff
@@ -1,8 +1,8 @@
 
 
-HEADERS = csm.h Error.h Warning.h Version.h Isd.h BytestreamIsd.h NitfIsd.h MultiNitfIsd.h Plugin.h Model.h GeometricModel.h RasterGM.h CorrelationModel.h FourParameterCorrelationModel.h LinearDecayCorrelationModel.h csmPointCloud.h PointCloudIsd.h PointCloudGM.h ModelIdentifier.h BundleGM.h Ellipsoid.h SettableEllipsoid.h
+HEADERS = csm.h Error.h Warning.h Version.h Isd.h BytestreamIsd.h NitfIsd.h MultiNitfIsd.h Plugin.h Model.h GeometricModel.h RasterGM.h SPDCorrelationFunction.h FourParameterCorrelationFunction.h LinearDecayCorrelationFunction.h DampedCosineCorrelationFunction.h CorrelationModel.h FourParameterCorrelationModel.h LinearDecayCorrelationModel.h MultiFunctionCorrelationModel.h csmPointCloud.h PointCl ...
```
This might be a good idea.. but I will leave it for now as there are other classes we will be adding soon, so this file will get touched again.

johnpdpkarp commented on FourParameterCorrelationFunction.cpp:
```diff
@@ -0,0 +1,156 @@
+{}
+
+void FourParameterCorrelationFunction::setCorrelationParameters(double a,
+                                                             double alpha,
```
done

roseej commented:
[requirement_brief_v2.pdf](https://github.com/ngageoint/csm/files/7847507/requirement_brief_v2.pdf)
This is the briefing which is scheduled to be presented on 1/12/22

johnpdpkarp commented:
I updated the files to provide a bit more accessibility to the parameters / parameter names.   I also added the ConstantCorrelationFunction to allow a constant correlation to be represented, since this is explicitly disallowed by the defined parameter ranges for the FourParameter Model.   (commit # 6)   Kevin Lam (BAE) found a bug in the updated LinearDecayCorrelationFunction class which I also fixed ( commit # 7)

johnpdpkarp commented:
Irwin Krinsky and Kevin Lam (MSP team) pointed out some inefficiencies with the design, stemming partially from populating the parameter names for each function...  Unfortunately, the names can't just be defined statically in all cases, since some, like the like the Linear Decay function, have a variable number of parameters.   I reworked the functions so that they parameter names / values aren't populated until asked for.  While this does still take time to do, it removes any of the associated computational inefficences, but it is doubtful someone would query for them often...   For most scenarios, all the user really needs is the computed correlation coefficient, which doesn't require the parameter names/values to be externally available.
